src/flamegraph/flamegraph.js

```diff
@@ -198,41 +199,42 @@ function update_text(e) {
 // zoom
 function zoom_reset(e) {
     if (e.attributes != undefined) {
-        orig_load(e, "x");
-        orig_load(e, "width");
+        if (e.attributes.x_samples != undefined) {
+            e.attributes.x.value = format_percent(100 * parseInt(e.attributes.x_samples.value) / total_samples);
+        }
+        if (e.attributes.width_samples != undefined) {
+            e.attributes.width.value = format_percent(100 * parseInt(e.attributes.width_samples.value) / total_samples);
+        }
     }
     if (e.childNodes == undefined) return;
     for(var i = 0, c = e.childNodes; i < c.length; i++) {
         zoom_reset(c[i]);
     }
 }
-function zoom_child(e, x, ratio) {
+function zoom_child(e, x, zoomed_width_samples) {
     if (e.attributes != undefined) {
         if (e.attributes.x != undefined) {
-            orig_save(e, "x");
-            e.attributes.x.value = format_percent((parseFloat(e.attributes.x.value) - x) * ratio);
             if (e.tagName == "text") {
                 e.attributes.x.value = format_percent(parseFloat(find_child(e.parentNode, "rect[x]").attributes.x.value) + (100 * 3 / frames.attributes.width.value));
```

Comment:
Should this also be updated to use `x_samples`?

Reply:
I'm not sure.

The percentage values for rectangles which are large enough to have text on them will be great enough that floating-point imprecision isn't significant. I think the code to do the computation would be wordy and not change the behavior.

Reply:
I'm more thinking that we might as well make all the computations consistent while we're at it, but happy to leave it either way.